## Taylor Swift Spotify Dataset - data preprocessing
The data used comes from [Kaggle](https://www.kaggle.com/datasets/jarredpriester/taylor-swift-spotify-dataset). This is a 
passion project used to learn the workflow and the rules of a proper analysis. The notebooks are based on previously written code, that is now cleaned up and structurized.

In [1]:
import pandas as pd
import numpy as np
from pathlib import Path

In [17]:
data_dir = Path("data")
input_path = Path("..") / data_dir / "taylor_swift_spotify.csv"

songs = pd.read_csv(input_path)
print("Shape of the data:")
print(songs.shape)
print()

Shape of the data:
(582, 18)



The songs count is 582 with 18 columns describing them. The final goal of this project is to create an ML model
that correctly predicts in which era (new Taylor vs old Taylor) the song was released, therefore we remove columns that do not serve an anlytical purpose,
such as realease_date or track_number. 

In [3]:
songs_info = songs.drop(['Unnamed: 0', 'release_date', 'track_number', 'id', 'uri'], axis=1)

Due to deluxe albums having extra content voice memos are present in the dataset. They are early versions
of songs that already exist in the dataset and have 0 popularity, so they will not be taken into consideration.

In [4]:
print(songs_info[songs_info['popularity'] == 0]['name'])
print()

songs_info = songs_info[songs_info['popularity'] != 0]
print("Shape after deleting the irrelevant columns and voice memos:")
print(songs_info.shape)

418       I Know Places - Voice Memo
419    I Wish You Would - Voice Memo
420         Blank Space - Voice Memo
Name: name, dtype: object

Shape after deleting the irrelevant columns and voice memos:
(579, 13)


For convenience we are changing full names of the albums to their common abbreviations. We do not take into consideration "Taylor's Versions" (TV) as they will be used later for experimentation.

In [34]:
album_full_names = ['THE TORTURED POETS DEPARTMENT: THE ANTHOLOGY', 
                    'Midnights (3am Edition)', 
                    'evermore (deluxe version)', 
                    'folklore (deluxe version)',
                    'Lover', 'reputation', 
                    '1989 (Deluxe)', 
                    'Red', 
                    'Speak Now', 
                    'Fearless (Platinum Edition)', 
                    'Taylor Swift (Deluxe Edition)',
                    
                    "1989 (Taylor's Version) [Deluxe]", 
                    "Speak Now (Taylor's Version)", 
                    "Red (Taylor's Version)", 
                    "Fearless (Taylor's Version)"]

songs_info = songs_info[songs_info['album'].isin(album_full_names)]

In [35]:
album_short_names = ['ttpd', 'midnights', 'evermore', 'folklore', 'lover', 'rep', '1989', 'red', 'speak now', 'fearless', 'debut', 
                     '1989', 'speak now', 'red', 'fearless']

album_dict = dict(zip(album_full_names, album_short_names))
songs_info['album'] = songs_info['album'].replace(album_dict)

In [7]:
songs_info.to_csv(f'../{data_dir}/taylor_songs_no_TV.csv', index=False)

### Column meaning

In [28]:
display(songs_info.sort_values('popularity', ascending=False).head())

,name,album,acousticness,danceability,energy,instrumentalness,liveness,loudness,speechiness,tempo,valence,popularity,duration_ms
324,Cruel Summer,lover,0.117,0.552,0.702,0.000021,0.1050,-5.707,0.1570,169.994,0.564,93,178426
325,Lover,lover,0.492,0.359,0.543,0.000016,0.1180,-7.582,0.0919,68.534,0.453,86,221306
344,Don’t Blame Me,rep,0.106,0.615,0.534,0.000018,0.0607,-6.719,0.0386,135.917,0.193,85,236413
12,I Can Do It With a Broken Heart,ttpd,0.123,0.701,0.751,0.000000,0.1500,-8.390,0.0502,129.994,0.220,82,218004
0,Fortnight (feat. Post Malone),ttpd,0.502,0.504,0.386,0.000015,0.0961,-10.976,0.0308,192.004,0.281,82,228965


**acousticness** - from 0 to 1, where 1 means most acoustic

**danceability** - from 0 to 1, where 1 means most danceable, takes into consideration tempo, rhythm stability, beat strength and regularity

**energy** - from 0 to 1, where 1 means most energetic, takes into consideration dynamic range, perceived loudness, timbre and entropy

**loudness** - The overall loudness of a track in decibels (dB). Values range between -60 and 0 db.

**speechiness** - detects the presence of spoken words in a track. The more exclusively speech-like the recording (e.g. talk show, audio book, poetry), the closer to 1.0 the attribute value. Values above 0.66 describe tracks that are probably made entirely of spoken words. Values between 0.33 and 0.66 describe tracks that may contain both music and speech, either in sections or layered, including such cases as rap music. Values below 0.33 most likely represent music and other non-speech-like tracks.

**tempo** - The overall estimated tempo of a track in beats per minute (BPM).

**valence** - A measure from 0.0 to 1.0 describing the musical positiveness conveyed by a track. Tracks with high valence sound more positive (e.g. happy, cheerful, euphoric), while tracks with low valence sound more negative (e.g. sad, depressed, angry).

### Storing relevant features for later use

In [39]:
albums = album_short_names

# Album categories
new_albums = albums[:6]

# Relevant features
features = ['acousticness', 'danceability', 'energy',
            'loudness', 'speechiness', 'tempo', 'valence']

In [40]:
%store albums
%store new_albums
%store features

Stored 'albums' (list)
Stored 'new_albums' (list)
Stored 'features' (list)
